gax-httpjson/src/main/java/com/google/api/gax/httpjson/ProtoRestSerializer.java

```diff
@@ -0,0 +1,149 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.httpjson;
+
+import com.google.api.core.BetaApi;
+import com.google.common.collect.ImmutableList;
+import com.google.protobuf.InvalidProtocolBufferException;
+import com.google.protobuf.Message;
+import com.google.protobuf.util.JsonFormat;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * This class is responsible for serializing/deserializing protobuf {@link Message} to/from a proper
+ * format suitable for transferring it over by in a HTTP REST request/response. The information from
+ * a {@code Message} is split between a request body (JSON), URL path parameters and query
+ * parameters.
+ */
+@BetaApi
+public class ProtoRestSerializer<RequestT extends Message> {
+  private ProtoRestSerializer() {}
+
+  /** Creates a new instance of ProtoRestSerializer. */
+  public static <RequestT extends Message> ProtoRestSerializer<RequestT> create() {
+    return new ProtoRestSerializer<>();
+  }
+
+  /**
+   * Serializes the data from {@code message} to a JSON string. The implementation relies on
+   * protobuf native JSON formatter.
+   *
+   * @param message a message to serialize
+   */
+  String toJson(RequestT message) {
+    try {
+      return JsonFormat.printer().print(message);
+    } catch (InvalidProtocolBufferException e) {
+      throw new RuntimeException("Failed to serialize message to JSON", e);
+    }
+  }
+
+  /**
+   * Deserializes a {@code message} from an input stream to a protobuf message.
+   *
+   * @param message the input stream with a JSON-encoded message in it
+   * @param builder an empty builder for the specific {@code RequestT} message to serialize
+   */
+  @SuppressWarnings("unchecked")
+  RequestT fromJson(InputStream message, Message.Builder builder) {
+    try (Reader json = new InputStreamReader(message)) {
+      JsonFormat.parser().ignoringUnknownFields().merge(json, builder);
+      return (RequestT) builder.build();
+    } catch (IOException e) {
+      throw new RuntimeException("Failed to parse response message", e);
```

Comment:
same story. Please, never do this.

Reply:
Replaced with a newly declared `ProtoRestSerializationException extends ApiException` (same as above) to stay consistent with the resot of the exceptions in gax-java and to avoid the the hassle of propagating/catching checked exceptions in the generated code.